src/IR/IR.hs

```diff
@@ -252,3 +259,117 @@ wellFormed (Prim p es _) = wfPrim p es && all wellFormed es
   wfPrim (PrimOp PrimLt    ) [_, _]    = True
   wfPrim (PrimOp PrimLe    ) [_, _]    = True
   wfPrim _                   _         = False
+
+
+{-
+how to use:
+1) stack ghci
+2) import IR.IR
+3) samples:
+show (LitIntegral 3)
+show New
+Questions:
+why does `show (Lit "a" Integer)` not work? -- b/c types not defined, just use () as a type
+why does `show (Var (VarId (Identifier "foo")) LitIntegral)` not work?
+
+show (Var (VarId (Identifier "foo")) ()))
+
+-}
+instance (Pretty t) => Show (Expr t) where
+  show p = show $ pretty p
+
+instance Show Primitive where
+  show p = show $ pretty p
+
+instance Show Literal where
+  show p = show $ pretty p
+
+instance (Pretty t) => Show (Program t) where
+  show p = show $ pretty p
+
+instance Pretty Literal where
+  pretty (LitIntegral i) = pretty (show i)
+  pretty (LitBool     b) = pretty (show b)
+
+instance Pretty Primitive where
+  pretty New        = pretty "new"
+  pretty Dup        = pretty "dup"
+  pretty Drop       = pretty "drop"
+  pretty Reuse      = pretty "reuse"
+  pretty Deref      = pretty "deref"
+  pretty Assign     = pretty "="
+  pretty After      = pretty "after"
+  pretty Fork       = pretty "fork"
+  pretty Wait       = pretty "wait"
+  pretty Loop       = pretty "loop"
+  pretty Break      = pretty "break"
+  pretty Return     = pretty "return"
+  pretty (PrimOp p) = pretty "PrimOp" <+> pretty p
+
+instance Pretty PrimOp where
+  pretty PrimNeg    = pretty "-"
+  pretty PrimNot    = pretty "!"
+  pretty PrimBitNot = pretty "~"
+  pretty PrimAdd    = pretty "+"
+  pretty PrimSub    = pretty "-"
+  pretty PrimMul    = pretty "*"
+  pretty PrimDiv    = pretty "/"
+  pretty PrimMod    = pretty "%"
+  pretty PrimBitAnd = pretty "&"
+  pretty PrimBitOr  = pretty "|"
+  pretty PrimEq     = pretty "=="
+  pretty PrimGt     = pretty ">"
+  pretty PrimGe     = pretty ">="
+  pretty PrimLt     = pretty "<"
+  pretty PrimLe     = pretty "<="
+
+instance (Pretty t) => Pretty (Expr t) where
+  {- do pretty show(VarId)?
+  this might be needed if other file uses legacy printer
+  var myInt LitIntegral?
+
+  //do things like varId take args?
+
+  //do i have to wrap each expression in parens somehow?
+  // this feels v. lispy
+  -}
+  pretty (Var a b) =
+    pretty "(var" <+> pretty (show a) <+> pretty b <+> pretty ")"
+  pretty (Data a b) =
+    pretty "(data" <+> pretty (show a) <+> pretty b <+> pretty ")"
+  pretty (Lit a b) = pretty "(literal" <+> pretty a <+> pretty b <+> pretty ")"
+  pretty (App a b c) =
+    pretty "(apply" <+> pretty a <+> pretty b <+> pretty c <+> pretty ")"
```

Comment:
this should parse into something like `(a) (b) : c`

Reply:
No and yes and no. 

No because the scanner is currently inserting braces in a way that breaks examples with matching parens. 

Yes because assuming the scanner is working, inserting parens defensively is the simplest way to get around precedence issues when "pretty"-printing. 

No because what you end up printing isn't actually very pretty, and there should be ways to use the pretty-printer monoid to only insert parens where needed.

Reply:
starting with parens seems safer - will do that for now

Reply:
I agree, fixing the frontend scanner seems like a me issue :)